net: add new options to `net.Socket` and `net.Server`

ShogunPanda opened the pull request:
Hello!
This PR adds the new following options to `net.Socket.connect`:

* `noDelay` (boolean):  If set to `true`, calls `uv_tcp_nodelay` on the socket after connection. By default it is disabled.
* `keepAlive` (boolean) and `keepAliveDelay` (number): If set to `true` calls `uv_tcp_nodelay` on the socket after connection. By default it is disabled.

This changes are needed as preliminary work on #34185 to avoid excessive JS to C++ bridging.
As I'm not a C++ expert, I encountered the following problems:

1. Not sure how to test changes (even though test suite is passing so apparently nothing is broken)
2. Error handling in `TCPWrap::AfterConnect` was tentative so some C++ expert should take a look. Moreover, what shall we do if `uv_tcp_connect` succeeds but any of `uv_tcp_nodelay` or `uv_tcp_keepalive` fails? Shall we close the socket or let user handle it? Currently I don't do anything.

Once this is initially reviewed, I will also add (either here or in a new PR) the analogous on inbound connections.

mcollina commented:
How do we test those options when set via the other methods?

addaleax commented:
> This changes are needed as preliminary work on #34185 to avoid excessive JS to C++ bridging.

I would want to avoid as much additional complexity as possible -- this adds quite a bit, and even adds new members to the TCPWrap class, even though they are only useful during connecting. I would personally prefer to not do this and instead call the existing C++ methods from JS.

> 2\. Moreover, what shall we do if `uv_tcp_connect` succeeds but any of `uv_tcp_nodelay` or `uv_tcp_keepalive` fails? Shall we close the socket or let user handle it?

Close the socket, yes, but also let the user handle the error itself.

addaleax commented on src/tcp_wrap.h:
```diff
@@ -64,6 +65,10 @@ class TCPWrap : public ConnectionWrap<TCPWrap, uv_tcp_t> {
 
+  bool noDelay;
+  bool keepAlive;
+  unsigned int keepAliveDelay;
```
**If** this PR adds new fields, then they should be using `lower_snake_case_;` as elsewhere in the code

ShogunPanda replied:
Sorry, I missed that. No worries, I'll fix once we agree on the final approach.

mcollina commented on src/tcp_wrap.cc:
```diff
@@ -325,14 +347,22 @@ void TCPWrap::Connect(const FunctionCallbackInfo<Value>& args,
   int err = uv_ip_addr(*ip_address, &addr);
 
+  wrap->noDelay = args[3]->IsTrue();
+  wrap->keepAlive = args[4]->IsTrue();
```
Can we attach the fields to the `ConnectWrap` instead?

In this way it will not increase the memory footprint for each connection.

ShogunPanda replied:
Sure, I can move them there so they are then discarded after connection.

@addaleax Any objection on using this approach? Or do you think calling JS methods will not impact performance too much?

addaleax replied:
@ShogunPanda Well yeah, @mcollina’s suggestion is definitely something that would improve this PR :slightly_smiling_face: 

> Or do you think calling JS methods will not impact performance too much?

I would still think that that’s absolutely preferable, and not expect significant performance overhead, because:

1. These calls are performed once-per-connection, which is much less than e.g. the number of data events for typical sockets, so it will have comparably little impact, and
2. Doing JS-only changes would still allow us to be just as fast in the common case (by calling `uv_tcp_nodelay(handle, 0);` in the connect handler, and only doing another call into C++ when the user explicitly requested that the value is *not* set)

Generally we try to stick to implementing as much logic in JS as possible, because it is the more accessible language of the two that Node.js is made out of :slightly_smiling_face:

ShogunPanda commented:
@addaleax  About testing, I just realized i can add two more testing in the array sent to `req_wrap->MakeCallback` in order to test it from the JS side. Do you think this will work?

addaleax commented:
> @addaleax About testing, I just realized i can add two more testing in the array sent to `req_wrap->MakeCallback` in order to test it from the JS side. Do you think this will work?

@ShogunPanda You mean, pass the values back to the callback in order to test that the values were passed along properly? I don’t think that’s necessary, and I don’t think we have other tests in that format. I think manual validation that the right syscalls are made (e.g. with `strace`) would be the only thing we can really do to test this here.

ShogunPanda commented:
> @ShogunPanda You mean, pass the values back to the callback in order to test that the values were passed along properly? I don’t think that’s necessary, and I don’t think we have other tests in that format. I think manual validation that the right syscalls are made (e.g. with `strace`) would be the only thing we can really do to test this here.

Ok, no problem at all. Will validate manually. I plan to resend this PR today for further evaluation.

ShogunPanda commented:
@addaleax @mcollina I've rewritten the PR in JS only (as I found out JS-to-CPP calls are not introducing any significant overhead). I also added some tests. 

PTAL, thanks!

mcollina commented on lib/net.js:
```diff
@@ -520,14 +534,18 @@ Socket.prototype.setNoDelay = function(enable) {
+Socket.prototype.setKeepAlive = function(enable, initialDelayMsecs) {
   if (!this._handle) {
-    this.once('connect', () => this.setKeepAlive(setting, msecs));
+    this.once('connect', () => this.setKeepAlive(enable, initialDelayMsecs));
```
You can avoid setting up this event listener now and just set `this[kSetKeepAlive]`, same for `kSetNoDelay`.

ShogunPanda replied:
Done!

mcollina reviewed (approved):
lgtm

ShogunPanda commented:
Any update on this? Shall I do anything else on this or are we good to merge?

Emill commented:
According to the documentation, these options have been added to the `connect(options)` method of `net.Socket`. However, according to the actual code, these options are added to the `net.Socket(options)` constructor. Hence, if you follow the documentation, these new options will not be used. This can be easily verified using `strace`.
